sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,729 @@
+semantics.
+
+[source,c++]
+----
+class context {
+
+  memory_pool context::ext_oneapi_get_default_memory_pool();
+
+  memory_pool context::ext_oneapi_get_default_memory_pool(device dev);
+
+}; // context
+----
+
+[source, c++]
+----
+memory_pool context::ext_oneapi_get_default_memory_pool();
+----
+
+_Returns_: The default memory pool associated with the context for allocating
+with the allocation kind `usm::host`.
+
+[source, c++]
+----
+memory_pool context::ext_oneapi_get_default_memory_pool(device dev);
+----
+
+_Returns_: The default memory pool associated with the context and `dev` for
+allocating with the allocation kind `usm::device`.
+
+
+=== Asynchronous malloc & free
+
+This extension introduces a series of new enqueue functions for enqueueing
+asynchornous malloc and free commands which operate with the memory pools also
+introduced in this extension.
+
+All enqueue functions introduced have overloads which take a SYCL `queue` and a
+SYCL `handler`. None of enqueue functions return a SYCL `event` direction, as
+this extension is in line with the
+link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
+  sycl_ext_oneapi_enqueue_functions] extension, so events are returned when
+calling `submit_with_event` and the `handler` overloads of these enqueue
+functions.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+void *async_malloc(queue q, size_t size);
+
+void *async_malloc(handler h, size_t size);
+
+void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);
+
+void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
+
+void async_free(queue q, void *ptr);
+
+void async_free(handler h, void *ptr);
+
+}  // ext::oneapi::experimental
+----
+
+[source, c++]
+----
+void *async_malloc(queue q, size_t size);
+
+void *async_malloc(handler h, size_t size);
+
+void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);
+
+void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
+----
+
+_Effects_: Enqueues a commands to `q` or the SYCL queue associated with `h`
+which will asynchornously allocate memory of size `size` in bytes, allocating
+from the memory pool `pool` if provided, otherwise allocation from the default
+memory pool associated with the SYCL context and device associated with `q` or
+`h`. Memory is first allocated from the memory pool if possible, otherwise
+memory is allocated from the OS to the memory pool to provide enough memory in
+the memory pool for the allocation.
+
+_Returns_: A pointer to the address of a memory reservation.
+
+_Throws_: An exception with the `errc::invalid` error code if `size` is zero.
+An exception with the `errc::memory_allocation` error code if the allocation
+brings the memory pool over the maximum size. This error can be thrown
+asynchornously.
+
+[_Note:_ Accessing the memory at the address of the pointer returned by
+asynchornous malloc functions before the command has completed execution is
```

Comment:
what does 'the command' represent?  the asynchronous malloc functions?

Reply:
So the command is a SYCL RT command that executes the asynchronous memory allocation from the memory pool, so until this command is complete is undefined behaviour to access the memory at that address.

Reply:
does this specific for out-of-order sycl queue?  If it is in-order sycl queue, I think there's no issue to first allocate memory async and then just access that memory address in a kernel.

Reply:
So the pointer can be passed to a kernel that is enqueued, but users must ensure that either the commands are either enqueued to the same in-order queue in which case, as you mentioned this will always be correct as the queue scheduling semantics guarantees correctness, or if enqueued to separate queues or an out-of-order queue the necessary dependencies are established.